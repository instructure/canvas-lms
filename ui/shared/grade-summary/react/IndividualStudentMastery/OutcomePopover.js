/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import _ from 'lodash'
import I18n from 'i18n!IndividualStudentMasteryOutcomePopover'
import {Flex} from '@instructure/ui-flex'
import {View} from '@instructure/ui-view'
import {Text} from '@instructure/ui-text'
import CalculationMethodContent from '../../backbone/models/CalculationMethodContent'
import {Popover} from '@instructure/ui-popover'
import {IconInfoLine} from '@instructure/ui-icons'
import DatetimeDisplay from '@canvas/datetime/react/components/DatetimeDisplay'
import {CloseButton, IconButton} from '@instructure/ui-buttons'
import {Modal} from '@instructure/ui-modal'
import WithBreakpoints, {breakpointsShape} from 'with-breakpoints'
import * as shapes from './shapes'

class OutcomePopover extends React.Component {
  static propTypes = {
    outcome: shapes.outcomeShape.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape,
    breakpoints: breakpointsShape
  }

  static defaultProps = {
    outcomeProficiency: null,
    breakpoints: {}
  }

  constructor() {
    super()
    this.state = {linkHover: false, linkClicked: false}
  }

  getSelectedRating() {
    const {outcomeProficiency} = this.props
    const {points_possible, mastery_points, score} = this.props.outcome
    const hasScore = score >= 0
    if (outcomeProficiency && hasScore) {
      const totalPoints = points_possible || mastery_points
      const percentage = totalPoints ? score / totalPoints : score
      const maxRating = outcomeProficiency.ratings[0].points
      const scaledScore = maxRating * percentage
      return (
        _.find(outcomeProficiency.ratings, r => scaledScore >= r.points) ||
        _.last(outcomeProficiency.ratings)
      )
    } else if (hasScore) {
      return _.find(this.defaultProficiency(mastery_points).ratings, r => score >= r.points)
    }
    return null
  }

  defaultProficiency = _.memoize(mastery_points => ({
    ratings: [
      {points: mastery_points * 1.5, color: '127A1B', description: I18n.t('Exceeds Mastery')},
      {points: mastery_points, color: '00AC18', description: I18n.t('Meets Mastery')},
      {points: mastery_points / 2, color: 'FAB901', description: I18n.t('Near Mastery')},
      {points: 0, color: 'EE0612', description: I18n.t('Well Below Mastery')}
    ]
  }))

  latestTime() {
    const {outcome} = this.props
    if (outcome.results.length > 0) {
      return _.sortBy(outcome.results, r => -r.submitted_or_assessed_at)[0].submitted_or_assessed_at
    }
    return null
  }

  renderSelectedRating() {
    const selectedRating = this.getSelectedRating()
    return (
      <Text size="small" weight="bold">
        <div>
          {selectedRating && (
            <div style={{color: `#${selectedRating.color}`}}>{selectedRating.description}</div>
          )}
        </div>
      </Text>
    )
  }

  renderPopoverContent() {
    const latestTime = this.latestTime()
    const popoverContent = new CalculationMethodContent(this.props.outcome).present()
    const {method, exampleText, exampleScores, exampleResult} = popoverContent
    const {outcome, breakpoints} = this.props
    const isVertical = !breakpoints.miniTablet

    return (
      <View as="div" padding="large" maxWidth="30rem">
        <CloseButton
          placement="end"
          onClick={() => this.setState({linkHover: false, linkClicked: false})}
        >
          {I18n.t('Click to close outcome details popover')}
        </CloseButton>
        <Text size="small">
          <Flex alignItems="stretch" direction="row" justifyItems="space-between">
            <Flex.Item grow shrink>
              {/* word-wrap used for IE support */}
              <div style={{wordWrap: 'break-word', overflowWrap: 'break-word'}}>
                {outcome.title}
              </div>
              <div>
                {isVertical && <div>{this.renderSelectedRating()}</div>}
                {I18n.t('Last Assessment: ')}
                {isVertical && <br />}
                {latestTime ? (
                  <DatetimeDisplay datetime={latestTime} format="%b %d, %l:%M %p" />
                ) : (
                  I18n.t('No submissions')
                )}
              </div>
              {outcome.friendly_description && (
                <div style={{padding: '0.5rem 0 0 0'}}>{outcome.friendly_description}</div>
              )}
            </Flex.Item>
            {!isVertical && <Flex.Item align="stretch">{this.renderSelectedRating()}</Flex.Item>}
          </Flex>
          <hr role="presentation" />
          <div>
            <Text size="small" weight="bold">
              {I18n.t('Calculation Method')}
            </Text>
            <div>{method}</div>
            <div style={{padding: '0.5rem 0 0 0'}}>
              <Text size="small" weight="bold">
                {I18n.t('Example')}
              </Text>
            </div>
            <div>{exampleText}</div>
            <div>{I18n.t('1- Item Scores: %{exampleScores}', {exampleScores})}</div>
            <div>{I18n.t('2- Final Score: %{exampleResult}', {exampleResult})}</div>
          </div>
        </Text>
      </View>
    )
  }

  renderPopover() {
    return (
      <span>
        <Popover
          show={this.state.linkHover || this.state.linkClicked}
          onDismiss={() => this.setState({linkHover: false, linkClicked: false})}
          placement="bottom"
          on={['hover', 'click']}
          shouldContainFocus
        >
          <Popover.Trigger>
            <IconButton
              size="small"
              margin="xx-small"
              withBackground={false}
              withBorder={false}
              screenReaderLabel={I18n.t('Click to expand outcome details')}
              renderIcon={IconInfoLine}
              onClick={() => this.setState(prevState => ({linkClicked: !prevState.linkClicked}))}
              onMouseEnter={() => this.setState({linkHover: true})}
              onMouseLeave={() => this.setState({linkHover: false})}
            />
          </Popover.Trigger>
          <Popover.Content>{this.renderPopoverContent()}</Popover.Content>
        </Popover>
      </span>
    )
  }

  renderModal() {
    return (
      <span>
        <IconButton
          size="small"
          margin="xx-small"
          withBackground={false}
          withBorder={false}
          screenReaderLabel={I18n.t('Click to expand outcome details')}
          renderIcon={IconInfoLine}
          onClick={() => this.setState(prevState => ({linkClicked: !prevState.linkClicked}))}
        />
        <Modal
          open={this.state.linkClicked}
          onDismiss={() =>
            this.setState(prevState => prevState.linkClicked && this.setState({linkClicked: false}))
          }
          size="fullscreen"
          label={I18n.t('Outcome Details')}
        >
          <Modal.Body>{this.renderPopoverContent()}</Modal.Body>
        </Modal>
      </span>
    )
  }

  render() {
    const {breakpoints} = this.props
    const modalLayout = !breakpoints.miniTablet
    if (modalLayout) {
      return this.renderModal()
    } else {
      return this.renderPopover()
    }
  }
}

export default WithBreakpoints(OutcomePopover)
