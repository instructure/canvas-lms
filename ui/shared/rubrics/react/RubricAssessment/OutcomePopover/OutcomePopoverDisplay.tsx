/*
 * Copyright (C) 2025 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {useScope as createI18nScope} from '@canvas/i18n'
import {Heading} from '@instructure/ui-heading'
import {Responsive} from '@instructure/ui-responsive'
import {View} from '@instructure/ui-view'
import {Flex} from '@instructure/ui-flex'
import {Text} from '@instructure/ui-text'
import CalculationMethodContent from '@canvas/grading/CalculationMethodContent'
import type {GetRubricOutcomeQuery} from '@canvas/graphql/codegen/graphql'
import OutcomeContextTag from '@canvas/outcome-context-tag'

const I18n = createI18nScope('rubrics-assessment-outcome')

type OutcomePopoverDisplayProps = {
  outcome: GetRubricOutcomeQuery['learningOutcome']
}
export const OutcomePopoverDisplay = ({outcome}: OutcomePopoverDisplayProps) => {
  if (!outcome) {
    return
  }

  const hasDisplayName = outcome.displayName && outcome.displayName.length > 0
  const displayName = hasDisplayName ? outcome.displayName : outcome.title

  const calculationMethodContent = new CalculationMethodContent({
    calculation_method: outcome.calculationMethod,
    calculation_int: outcome.calculationInt,
    is_individual_outcome: true,
    mastery_points: outcome.masteryPoints,
  }).present()

  const {friendlyCalculationMethod, exampleText} = calculationMethodContent ?? {}

  return (
    <Responsive
      match="media"
      query={{
        compact: {maxWidth: '50rem'},
        fullWidth: {minWidth: '50rem'},
        large: {minWidth: '66.5rem'},
      }}
    >
      {(_props, matches) => {
        const isFullWidth = matches?.includes('fullWidth') ?? false
        const maxWidth = isFullWidth ? '600px' : '300px'

        return (
          <View as="div" display="block" data-testid="outcome-popover-display" maxWidth={maxWidth}>
            <Heading
              level="h3"
              as="h3"
              margin="x-small 0"
              data-testid="outcome-popover-display-name"
            >
              {displayName}
            </Heading>
            {hasDisplayName && (
              <Heading level="h4" as="h4" margin="x-small 0" data-testid="outcome-popover-title">
                {outcome.title}
              </Heading>
            )}
            <View
              as="div"
              data-testid="outcome-popover-display-content-description"
              dangerouslySetInnerHTML={{__html: outcome.description ?? ''}}
            />
            <OutcomeContextTag
              outcomeContextType={outcome.contextType ?? undefined}
              outcomeContextId={outcome.contextId ?? undefined}
              margin="0 0 x-small 0"
            />
            {friendlyCalculationMethod && exampleText && (
              <>
                <View as="hr" />
                <Flex as="div" width="100%" gap="medium" wrap={isFullWidth ? 'no-wrap' : 'wrap'}>
                  <Flex.Item align="start">
                    <Heading level="h4" as="h4">
                      {I18n.t('Calculation Method')}
                    </Heading>

                    <Text data-testid="outcome-popover-display-content-calculation-method">
                      {friendlyCalculationMethod}
                    </Text>
                  </Flex.Item>
                  <Flex.Item shouldGrow shouldShrink align="start">
                    <View as="div">
                      <Heading level="h4" as="h4">
                        {I18n.t('Example')}
                      </Heading>
                      <View as="div">
                        <Text
                          wrap="break-word"
                          data-testid="outcome-popover-display-content-example"
                        >
                          {exampleText}
                        </Text>
                      </View>
                    </View>
                  </Flex.Item>
                </Flex>
              </>
            )}
          </View>
        )
      }}
    </Responsive>
  )
}
