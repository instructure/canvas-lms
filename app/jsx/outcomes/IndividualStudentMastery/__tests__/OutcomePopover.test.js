/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {shallow} from 'enzyme'
import OutcomePopover from '../OutcomePopover'

const time1 = new Date(Date.UTC(2018, 1, 1, 7, 1, 0))
const time2 = new Date(Date.UTC(2018, 1, 1, 8, 1, 0))

const defaultProps = (props = {}) => ({
  outcome: {
    id: 1,
    assignments: [],
    expansionId: 100,
    mastered: false,
    mastery_points: 3,
    points_possible: 5,
    calculation_method: 'highest',
    score: 3,
    ratings: [{description: 'My first rating'}, {description: 'My second rating'}],
    results: [
      {
        id: 1,
        score: 1,
        percent: 0.1,
        assignment: {
          id: 1,
          html_url: 'http://foo',
          name: 'My assignment',
          submission_types: 'online_quiz',
          score: 0
        },
        submitted_or_assessed_at: time1
      },
      {
        id: 1,
        score: 7,
        percent: 0.7,
        assignment: {
          id: 2,
          html_url: 'http://bar',
          name: 'Assignment 2',
          submission_types: 'online_quiz',
          score: 3
        },
        submitted_or_assessed_at: time2
      }
    ],
    title: 'My outcome'
  },
  outcomeProficiency: {
    ratings: [
      {color: 'blue', description: 'I am blue', points: 10},
      {color: 'green', description: 'I am Groot', points: 5},
      {color: 'red', description: 'I am red', points: 0}
    ]
  },
  ...props
})

it('renders the OutcomePopover component', () => {
  const wrapper = shallow(<OutcomePopover {...defaultProps()} />)
  expect(wrapper).toMatchSnapshot()
})

it('renders correctly with no results', () => {
  const props = defaultProps()
  props.outcome.results = []
  const wrapper = shallow(<OutcomePopover {...props} />)
  expect(wrapper).toMatchSnapshot()
})

it('renders correctly with no custom outcomeProficiency', () => {
  const props = defaultProps()
  props.outcomeProficiency = null
  const wrapper = shallow(<OutcomePopover {...props} />)
  expect(wrapper).toMatchSnapshot()
})

it('properly expands details for screenreader users', () => {
  const props = defaultProps()
  const wrapper = shallow(<OutcomePopover {...props} />)
  expect(wrapper.state('linkClicked')).toEqual(false)
  wrapper.find('Link').simulate('click')
  expect(wrapper.state('linkClicked')).toEqual(true)
})

describe('latestTime', () => {
  it('properly returns the most recent submission time', () => {
    const props = defaultProps()
    const wrapper = shallow(<OutcomePopover {...props} />)
    expect(wrapper.instance().latestTime()).toEqual(time1)
  })

  it('properly returns nothing when there are no results', () => {
    const props = defaultProps()
    props.outcome.results = []
    const wrapper = shallow(<OutcomePopover {...props} />)
    expect(wrapper.instance().latestTime()).toBeNull()
  })
})

describe('getSelectedRating', () => {
  it('properly returns the custom proficiency level', () => {
    const props = defaultProps()
    const wrapper = shallow(<OutcomePopover {...props} />)
    const rating = wrapper.instance().getSelectedRating()
    expect(rating.description).toEqual('I am Groot')
  })

  it('properly returns the default proficiency level', () => {
    const props = defaultProps()
    props.outcomeProficiency = null
    const wrapper = shallow(<OutcomePopover {...props} />)
    const rating = wrapper.instance().getSelectedRating()
    expect(rating.description).toEqual('Meets Mastery')
  })
})
