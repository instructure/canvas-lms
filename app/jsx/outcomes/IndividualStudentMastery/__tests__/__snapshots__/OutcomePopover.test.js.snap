// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no custom outcomeProficiency 1`] = `
<span>
  <Popover
    alignArrow={false}
    constrain="window"
    contentRef={[Function]}
    defaultFocusElement={null}
    defaultShow={false}
    insertAt="bottom"
    label={null}
    liveRegion={null}
    mountNode={null}
    offsetX={0}
    offsetY={0}
    on={
      Array [
        "hover",
        "click",
      ]
    }
    onBlur={[Function]}
    onClick={[Function]}
    onDismiss={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    onPositionChanged={[Function]}
    onPositioned={[Function]}
    onShow={[Function]}
    onToggle={[Function]}
    placement="bottom"
    positionTarget={null}
    shadow="resting"
    shouldCloseOnDocumentClick={true}
    shouldCloseOnEscape={true}
    shouldContainFocus={true}
    shouldFocusContentOnTriggerBlur={false}
    shouldRenderOffscreen={false}
    shouldReturnFocus={true}
    show={false}
    stacking="topmost"
    trackPosition={true}
    variant="default"
    withArrow={true}
  >
    <PopoverTrigger>
      <Link
        disabled={false}
        iconPlacement="start"
        linkRef={[Function]}
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        variant="default"
      >
        <span
          style={
            Object {
              "color": "black",
            }
          }
        >
          <IconInfoLine />
        </span>
        <span>
          <ScreenReaderContent
            as="span"
          >
            Click to expand outcome details
          </ScreenReaderContent>
        </span>
      </Link>
    </PopoverTrigger>
    <PopoverContent>
      <View
        as="div"
        borderColor="default"
        debug={false}
        display="auto"
        focusColor="info"
        focusPosition="offset"
        focused={false}
        maxWidth="30rem"
        overflowX="visible"
        overflowY="visible"
        padding="large"
        position="static"
        shouldAnimateFocus={true}
      >
        <CloseButton
          buttonRef={[Function]}
          margin="0"
          offset="x-small"
          onClick={[Function]}
          placement="end"
          size="small"
          variant="icon"
        >
          Click to close outcome details popover
        </CloseButton>
        <Text
          as="span"
          letterSpacing="normal"
          size="small"
          wrap="normal"
        >
          <Flex
            alignItems="stretch"
            as="span"
            direction="row"
            elementRef={[Function]}
            inline={false}
            justifyItems="space-between"
            visualDebug={false}
            wrapItems={false}
          >
            <FlexItem
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <div
                style={
                  Object {
                    "overflowWrap": "break-word",
                    "wordWrap": "break-word",
                  }
                }
              >
                My outcome
              </div>
              <div>
                Last Assessment: 
                <DatetimeDisplay
                  datetime={2018-02-01T07:01:00.000Z}
                  format="%b %d, %l:%M %p"
                />
              </div>
            </FlexItem>
            <FlexItem
              align="stretch"
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                <div>
                  <div
                    style={
                      Object {
                        "color": "#00AC18",
                        "textAlign": "end",
                      }
                    }
                  >
                    Meets Mastery
                  </div>
                </div>
              </Text>
            </FlexItem>
          </Flex>
          <hr
            role="presentation"
          />
          <div>
            <Text
              as="span"
              letterSpacing="normal"
              size="small"
              weight="bold"
              wrap="normal"
            >
              Calculation Method
            </Text>
            <div>
              Highest Score
            </div>
            <div
              style={
                Object {
                  "padding": "0.5rem 0 0 0",
                }
              }
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                Example
              </Text>
            </div>
            <div>
              Mastery score reflects the highest score of a graded assignment or quiz.
            </div>
            <div>
              1- Item Scores: 1, 4, 2, 3
            </div>
            <div>
              2- Final Score: 4
            </div>
          </div>
        </Text>
      </View>
    </PopoverContent>
  </Popover>
</span>
`;

exports[`renders correctly with no results 1`] = `
<span>
  <Popover
    alignArrow={false}
    constrain="window"
    contentRef={[Function]}
    defaultFocusElement={null}
    defaultShow={false}
    insertAt="bottom"
    label={null}
    liveRegion={null}
    mountNode={null}
    offsetX={0}
    offsetY={0}
    on={
      Array [
        "hover",
        "click",
      ]
    }
    onBlur={[Function]}
    onClick={[Function]}
    onDismiss={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    onPositionChanged={[Function]}
    onPositioned={[Function]}
    onShow={[Function]}
    onToggle={[Function]}
    placement="bottom"
    positionTarget={null}
    shadow="resting"
    shouldCloseOnDocumentClick={true}
    shouldCloseOnEscape={true}
    shouldContainFocus={true}
    shouldFocusContentOnTriggerBlur={false}
    shouldRenderOffscreen={false}
    shouldReturnFocus={true}
    show={false}
    stacking="topmost"
    trackPosition={true}
    variant="default"
    withArrow={true}
  >
    <PopoverTrigger>
      <Link
        disabled={false}
        iconPlacement="start"
        linkRef={[Function]}
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        variant="default"
      >
        <span
          style={
            Object {
              "color": "black",
            }
          }
        >
          <IconInfoLine />
        </span>
        <span>
          <ScreenReaderContent
            as="span"
          >
            Click to expand outcome details
          </ScreenReaderContent>
        </span>
      </Link>
    </PopoverTrigger>
    <PopoverContent>
      <View
        as="div"
        borderColor="default"
        debug={false}
        display="auto"
        focusColor="info"
        focusPosition="offset"
        focused={false}
        maxWidth="30rem"
        overflowX="visible"
        overflowY="visible"
        padding="large"
        position="static"
        shouldAnimateFocus={true}
      >
        <CloseButton
          buttonRef={[Function]}
          margin="0"
          offset="x-small"
          onClick={[Function]}
          placement="end"
          size="small"
          variant="icon"
        >
          Click to close outcome details popover
        </CloseButton>
        <Text
          as="span"
          letterSpacing="normal"
          size="small"
          wrap="normal"
        >
          <Flex
            alignItems="stretch"
            as="span"
            direction="row"
            elementRef={[Function]}
            inline={false}
            justifyItems="space-between"
            visualDebug={false}
            wrapItems={false}
          >
            <FlexItem
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <div
                style={
                  Object {
                    "overflowWrap": "break-word",
                    "wordWrap": "break-word",
                  }
                }
              >
                My outcome
              </div>
              <div>
                Last Assessment: 
                No submissions
              </div>
            </FlexItem>
            <FlexItem
              align="stretch"
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                <div>
                  <div
                    style={
                      Object {
                        "color": "#green",
                        "textAlign": "end",
                      }
                    }
                  >
                    I am Groot
                  </div>
                </div>
              </Text>
            </FlexItem>
          </Flex>
          <hr
            role="presentation"
          />
          <div>
            <Text
              as="span"
              letterSpacing="normal"
              size="small"
              weight="bold"
              wrap="normal"
            >
              Calculation Method
            </Text>
            <div>
              Highest Score
            </div>
            <div
              style={
                Object {
                  "padding": "0.5rem 0 0 0",
                }
              }
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                Example
              </Text>
            </div>
            <div>
              Mastery score reflects the highest score of a graded assignment or quiz.
            </div>
            <div>
              1- Item Scores: 1, 4, 2, 3
            </div>
            <div>
              2- Final Score: 4
            </div>
          </div>
        </Text>
      </View>
    </PopoverContent>
  </Popover>
</span>
`;

exports[`renders the OutcomePopover component 1`] = `
<span>
  <Popover
    alignArrow={false}
    constrain="window"
    contentRef={[Function]}
    defaultFocusElement={null}
    defaultShow={false}
    insertAt="bottom"
    label={null}
    liveRegion={null}
    mountNode={null}
    offsetX={0}
    offsetY={0}
    on={
      Array [
        "hover",
        "click",
      ]
    }
    onBlur={[Function]}
    onClick={[Function]}
    onDismiss={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    onPositionChanged={[Function]}
    onPositioned={[Function]}
    onShow={[Function]}
    onToggle={[Function]}
    placement="bottom"
    positionTarget={null}
    shadow="resting"
    shouldCloseOnDocumentClick={true}
    shouldCloseOnEscape={true}
    shouldContainFocus={true}
    shouldFocusContentOnTriggerBlur={false}
    shouldRenderOffscreen={false}
    shouldReturnFocus={true}
    show={false}
    stacking="topmost"
    trackPosition={true}
    variant="default"
    withArrow={true}
  >
    <PopoverTrigger>
      <Link
        disabled={false}
        iconPlacement="start"
        linkRef={[Function]}
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        variant="default"
      >
        <span
          style={
            Object {
              "color": "black",
            }
          }
        >
          <IconInfoLine />
        </span>
        <span>
          <ScreenReaderContent
            as="span"
          >
            Click to expand outcome details
          </ScreenReaderContent>
        </span>
      </Link>
    </PopoverTrigger>
    <PopoverContent>
      <View
        as="div"
        borderColor="default"
        debug={false}
        display="auto"
        focusColor="info"
        focusPosition="offset"
        focused={false}
        maxWidth="30rem"
        overflowX="visible"
        overflowY="visible"
        padding="large"
        position="static"
        shouldAnimateFocus={true}
      >
        <CloseButton
          buttonRef={[Function]}
          margin="0"
          offset="x-small"
          onClick={[Function]}
          placement="end"
          size="small"
          variant="icon"
        >
          Click to close outcome details popover
        </CloseButton>
        <Text
          as="span"
          letterSpacing="normal"
          size="small"
          wrap="normal"
        >
          <Flex
            alignItems="stretch"
            as="span"
            direction="row"
            elementRef={[Function]}
            inline={false}
            justifyItems="space-between"
            visualDebug={false}
            wrapItems={false}
          >
            <FlexItem
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <div
                style={
                  Object {
                    "overflowWrap": "break-word",
                    "wordWrap": "break-word",
                  }
                }
              >
                My outcome
              </div>
              <div>
                Last Assessment: 
                <DatetimeDisplay
                  datetime={2018-02-01T07:01:00.000Z}
                  format="%b %d, %l:%M %p"
                />
              </div>
            </FlexItem>
            <FlexItem
              align="stretch"
              as="span"
              elementRef={[Function]}
              grow={true}
              shrink={true}
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                <div>
                  <div
                    style={
                      Object {
                        "color": "#green",
                        "textAlign": "end",
                      }
                    }
                  >
                    I am Groot
                  </div>
                </div>
              </Text>
            </FlexItem>
          </Flex>
          <hr
            role="presentation"
          />
          <div>
            <Text
              as="span"
              letterSpacing="normal"
              size="small"
              weight="bold"
              wrap="normal"
            >
              Calculation Method
            </Text>
            <div>
              Highest Score
            </div>
            <div
              style={
                Object {
                  "padding": "0.5rem 0 0 0",
                }
              }
            >
              <Text
                as="span"
                letterSpacing="normal"
                size="small"
                weight="bold"
                wrap="normal"
              >
                Example
              </Text>
            </div>
            <div>
              Mastery score reflects the highest score of a graded assignment or quiz.
            </div>
            <div>
              1- Item Scores: 1, 4, 2, 3
            </div>
            <div>
              2- Final Score: 4
            </div>
          </div>
        </Text>
      </View>
    </PopoverContent>
  </Popover>
</span>
`;
